tests/model/schema.js

```diff
@@ -1116,112 +1115,145 @@ describe( 'Schema', () => {
 			root = doc.createRoot();
 
 			schema.register( 'p', { inheritAllFrom: '$block' } );
-			schema.register( 'h1', { inheritAllFrom: '$block' } );
-			schema.register( 'img', {
-				allowWhere: '$text'
-			} );
+			schema.register( 'img', { allowWhere: '$text' } );
 
-			schema.addAttributeCheck( ( ctx, attributeName ) => {
-				// Allow 'bold' on p>$text.
-				if ( ctx.endsWith( 'p $text' ) && attributeName == 'bold' ) {
-					return true;
-				}
+			// This is a "hack" to allow setting any ranges in `setData` util.
+			schema.extend( '$text', { allowIn: '$root' } );
+		} );
 
-				// Allow 'bold' on $root>p.
-				if ( ctx.endsWith( '$root p' ) && attributeName == 'bold' ) {
-					return true;
-				}
-			} );
+		function test( input, attribute, output ) {
```

Comment:
Please don't. Your meaningless `test`function wrap 4 lines of code and then you use 5 lines of code to execute it. I had to check what `foo` (the second parameter) is to ensure what is going on in tests. Instead you can wrap:

```
const sel = new Selection( validRanges );
expect( stringify( root, sel ) ).to.equal( output );
```

into a simple util or make `stringify` accept an array of ranges.

Reply:
I honestly don't agree and I really insist on keeping it the way it is. The function makes code cleaner and test cases easy to look and inspect. For example, instead of:

```javascript
test(
	'[<p>foo<img></img>bar</p>]',
	'src',
	'<p>foo[<img></img>]bar</p>'
);
```

You'd have:

```javascript
setData( model, '[<p>foo<img></img>bar</p>]' );
const validRanges = schema.getValidRanges( doc.selection.getRanges(), 'src' );

expectUtil( validRanges, '<p>foo[<img></img>]bar</p>' );
```

Which looks much worse.

It doesn't matter if I "use 5 lines to execute 4 lines" cause it's all about how clear the code looks like. I could easily write it as:

```javascript
test( '[<p>foo<img></img>bar</p>]', 'foo', '[<p>foo<img></img>bar</p>]' );
```

If I wanted but it doesn't make code easier to analyze.

So, to sum up, I strongly disagree with your comment. You wrote that you don't like something but your arguments aren't convincing.

> I had to check what foo (the second parameter) is to ensure what is going on in tests.

Alright, but after you checked (and it's not like a rocket science and can even be deduced from the context) you know it and all other cases are easier to read. Mind you, you would have to understand the first scenario no matter if there would be a test helper or not.

Reply:
Ok, at least change the name of the function to `assertValidRangesForAttribute`.